### ISSUE TITLE
Delete chat message

### ISSUE THREAD

**Issue by edwh** [Freegle/iznik-nuxt3#2]
If you click on a chat message that you've sent, there's an option to `Mark unread`.  We need to:

1. Add an option to `Delete`
2. Create a confirmation modal - similar to `ChatHideModal`, which asks them to confirm this.  We need to tell them that we will delete it from our systems, but it's possible that the other freegler may have received the message by email (which we can't delete).
3. Add a `delete` action into the `chat.js` store.
4. Add a `delete` call into `ChatAPI.js` which makes a v1 API `$del` request.
5. Shows a confirmation or the error.

The server does not yet handle that verb, so it will return an error until it does - but that's one of the cases to test.

**Comment by edwh** [comment]
@smellyshovel The server API call now exists; send a DELETE to `/chatmessages` with a parameter of `id` and the chat message id.

**Comment by smellyshovel** [comment]
@edwh Should it be possible to delete other chat participant's messages?

**Comment by smellyshovel** [comment]
@edwh another question. Where should I specify the chat's and message's IDs? `DELETE` requests typically do not support payloads, so I guess those should be added to the query params? What should be their names?

**Comment by smellyshovel** [comment]
What's the point of calling `$postOverride` with the `overrideMethod` parameter when it's ignored and the actual method is intead always set to be "POST"? Is there something that i'm missing? Or is it just a bug/typo?

![Screenshot_20230818_135247](https://github.com/Freegle/iznik-nuxt3/assets/9782236/7021d002-fa13-4026-b172-aae758b15fea)

**Comment by edwh** [comment]
> @edwh Should it be possible to delete other chat participant's messages?

No.

**Comment by edwh** [comment]
> @edwh another question. Where should I specify the chat's and message's IDs? `DELETE` requests typically do not support payloads, so I guess those should be added to the query params? What should be their names?

A single query parameter of `id` with the chat message id.  The id is globally unique, so the server can tell whether or not it is in one of the user's chats.

**Comment by edwh** [comment]
> What's the point of calling `$postOverride` with the `overrideMethod` parameter when it's ignored and the actual method is intead always set to be "POST"? Is there something that i'm missing? Or is it just a bug/typo?

Good question.  We found from experience a few years ago that there are some unusual network environments which block HTTP verbs like PUT and DELETE, but allow POSTs.  https://stackoverflow.com/questions/1828790/restful-put-and-delete-and-firewalls has some background on this.

It's unclear whether this is still an issue nowadays, but we don't really want to find out.  So we treat the other verbs by adding the `X-HTTP-Method-Override` header, and then the server detects that and processes it as though it had been the corresponding HTTP verb.

**Comment by smellyshovel** [comment]
@edwh 

Request:
```
https://fdapilive.ilovefreegle.org/api/chatmessage?id=77381962&loggedInAs=43283758&requestid=13
POST (X-Http-Method-Override: DELETE)
```

Response:
```
{
  "ret": 1000,
  "status": "Invalid API call",
  "call": "chatmessage",
  "type": "DELETE",
  "session": "",
  "duration": 0.004227876663208008,
  "cpucost": 0.0004300446736268724,
  "dbwaittime": 0.004372119903564453,
  "includetime": 5.0067901611328125e-5
}
```

Is that expected?

**Comment by edwh** [comment]
I think it's chatmessages with an s.

**Comment by smellyshovel** [comment]
> I think it's chatmessages with an s.

Indeed. Thanks.

So, for the most part it's ready, though:
1. I don't see that the message content is updated to be `(Message deleted)` even though I call the `fetchMessages` method again after the message has been deleted. Most probably that's due to the fact that the existing content update logic is not supported (there's no edit-message-kind-of functionality, is there?). So when I call `fetchMessages`, it won't update the contents of the existing ones. Though that's just my guess, haven't taken a look at the implementation yet. Can you confirm that it might be the case?
2. Could you please give me a name of some component which showcases the 5th requirement: showing error and success messages, so that I could take it as an example? Thanks!

**Comment by edwh** [comment]
1. Does it return that from the server in the network, or is the old text still there?  If it's a server bug I can have a look,
2. Have a look at `ForgetFailModal` for an example of a failure modal; use this as a template but add a prop to indicate whether it was a success or failure.

**Comment by smellyshovel** [comment]
> 1. Does it return that from the server in the network, or is the old text still there?  If it's a server bug I can have a look,
> 2. Have a look at `ForgetFailModal` for an example of a failure modal; use this as a template but add a prop to indicate whether it was a success or failure.

1. The server successfully returns the updated messages. But the patches are not applied on the client side. So, the question is that maybe you know if that's expectable (like if the logic of fetchMessages and its underlying calls wasn't implemented with such updates in mind) or if that's something that I miss on my end.
2. Aha, so you want it to be a modal, Ok. I though of some sort of a toast instead. But OK, that's clear, thanks for the tip!

**Comment by smellyshovel** [comment]
Ok, the question is no more. Found this comment in the `fetchMessages` implementation:

```
    async fetchMessages(id) {
      const messages = await api(this.config).chat.fetchMessages(id)

      // Update the store with care - we want to pick up new/changed messages but we don't want to trigger
      // unnecessary reactivity updates if we have fetched a message that is already in the cache.
```

**Comment by edwh** [comment]
@smellyshovel Yes, I was just looking.  So because the number of chat messages isn't the same, it will skip the update.  You can add a "force" flag.

**Comment by smellyshovel** [comment]
@edwh OK, I'll do that, but it's the same story as with myposts. Though there we're restricted mostly by the number of simultaneous API calls, and here we don't have them. But we do have infinite scrolling, the the total number of loaded last messages rarely exceeds 12-15. What we're doing is considered axhaustive micro-optimizations, which is not cool. Vue can handle those updates without any impact on the performance for sure. But yeah, I got you, will implement with a flag.

**Comment by edwh** [comment]
I didn't add that code just for fun!  I added later it because I was doing performance optimisation on stuff I could see was actually noticeably slow.
